### PR TITLE
feat(dex): integration mdex

### LINKED ISSUE #5796: [Bounty] Integrate MDEX, A Trading Platform used for AMM Technology for Swap Pools into Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks 
- Bounty Size
  - [x] 1000 USDC or $MASK equivalent 

## Part 1: Problem Description
**Background** MDEX, an acronym to Mandala Exchange, is a leading decentralized exchange protocol built on the Huobi chain. A trading platform using the automated market maker technology for fund pools.

It is part of MDEX plan to build a creative DEX, DAO, and IMO/ICO on ETH and Heco. This is to provide a configuration and asset selection that is more reliable and safe for users.

It uses a mixed or dual mechanism in its mining operations which are, transaction and liquidity mechanisms. Similar to other Cryptocurrencies, MDEX tokens (MDX) can be used for various purposes, including; serving as a medium for trading, voting, repurchase, and fundraising, among others.

**Solution** Create a plugin that integrates  MDEX on BSC and ETH into Mask. The UI should be identical to other swaps. 

image

image

_MDEX Swap Page_
Screen Shot 2022-03-06 at 6 14 00 PM

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://mdexdoc.gitbook.io/doc/
- https://github.com/mdexSwap
- https://bsc.mdex.com/#/swap?lang=en

## Part 3: Contact and Support
@yisiliu @tzhan28 @LindaMask for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q,
MDEX Discord: https://discord.com/invite/3TYDPktjqC

Please speak up in the Mask discord's bounty section or just discord dm @yisiliu or @taylor zhang or @lin_dao #5796  so that we can find you on Discord. For faster communication reach out to lin_DAO on telegram and she will create a dedicated group with the Mask team to start the bounty process

Submission Requirements
Please include demo to show the integration is working.

Judging Criteria
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by developerfred**
## Description

Closes  #5796

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [] I have commented on my code, particularly in hard-to-understand areas.
- [] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by developerfred** [comment]
@wenluomask

File: packages/web3-constants/evm/token.json [hunk 1]
```diff
@@ -899,6 +899,26 @@
         "Aurora": "",
         "Aurora_Testnet": ""
     },
+    "MDX_ADDRESS": {
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token.json]: What does this field for?

File: packages/mask/src/plugins/Trader/constants/trader.ts [hunk 2]
```diff
@@ -50,6 +50,7 @@ export const CUSD = createERC20Tokens('cUSD_ADDRESS', 'Celo Dollar', 'cUSD', 18)
 export const CEUR = createERC20Tokens('cEUR_ADDRESS', 'Celo Euro', 'cEUR', 18)
 export const JOE = createERC20Tokens('JOE_ADDRESS', 'JoeToken', 'JOE', 18)
 export const PNG = createERC20Tokens('PNG_ADDRESS', 'Pangolin', 'PNG', 18)
+export const MDX = createERC20Tokens('MDX_ADDRESS', ' MDX Token', 'MDX', 18)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/constants/trader.ts]: It's not used.

File: packages/web3-constants/evm/token.json (hunk 1 above)
Comment (developerfred) [reply packages/web3-constants/evm/token.json]: I will delete it, it will be necessary when heco was implemented.

**Review by Jack-Works — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

File: packages/web3-constants/evm/trader.json [hunk 3]
```diff
@@ -1018,5 +1018,85 @@
...
+    },
+    "MDEX_FACTORY_ADDRESS": {
+        "Mainnet": "0x7DAe51BD3E3376B8c7c4900E9107f12Be3AF1bA8",
+        "Ropsten": "",
+        "Rinkeby": "",
+        "Kovan": "",
+        "Gorli": "",
+        "BSC": "0x3CD1C46068dAEa5Ebb0d3f55F6915B10648062B8",
+        "BSCT": "",
+        "Matic": "",
```
Comment (UncleBill) [inline packages/web3-constants/evm/trader.json:1050]: Could you provide the source of these contract address? Including the `INIT_CODE_HASH`.  Where do you get them from? We need to verify them to make sure they are official contracts.

Comment (developerfred) [reply packages/web3-constants/evm/trader.json:1050]: Heco Network -> https://github.com/mdexSwap/contracts
Binance -> https://github.com/mdexSwap/bscswap

Comment (UncleBill) [reply packages/web3-constants/evm/trader.json:1050]: I don't have Heco yet.
I have checked https://github.com/mdexSwap/bscswap for BSC too. What about for Mainnet?

Comment (developerfred) [reply packages/web3-constants/evm/trader.json:1050]: @UncleBill check on deployer: https://etherscan.io/address/0xc6f3b19075f20ad4e9efc7662ac2631d90aa7226
for binance and mainnet is the same implementer

Comment (UncleBill) [reply packages/web3-constants/evm/trader.json:1050]: OK. I can verify them now.

**Comment by UncleBill** [comment]
@DimensionDev/qa Please help to verify.

**Review by UncleBill — DISMISSED** [review]

**Comment by wenluomask** [comment]
Please merge latest develop branch into this pr. Currently swap button is unresponsive in Dashboard Wallets page and token list cannot open in Popup's swap page.

Screen Shot 2022-03-18 at 11 44 12 PM
Screen Shot 2022-03-18 at 11 44 23 PM

**Comment by developerfred** [comment]
@wenluomask done

**Comment by wenluomask** [comment]
MDEX in BNB Chain and ETH Checked.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**